### PR TITLE
Update systemd_journal parser

### LINKED ISSUE #4728: systemd journal parser does not parse logs from systemd 249
**Describe the problem:**

Please provide a clear and detailed description of what the problem is.

**To Reproduce:**

The version of Plaso you used:

20230623

The operating system you are running Plaso on (Not the operating system of the image/files you're trying to analyze):

Ubuntu 22.04

```
systemd --version                 
systemd 249 (249.11-0ubuntu3.9)
+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
```

Steps to reproduce the behavior including command line and arguments and output:

Running:

```./tools/log2timeline.py --storage_file ~/test.plaso /var/log/journal/```

results in only filestat events (no systemd_journal events).

**Expected behavior:**

I would expect there to be systemd_journal events (that correspond to journalctl entries).

**Additional context**

Any other context about the problem here.

### PR DESCRIPTION
**Pull request by michael-ashburn**
## One line description of pull request
Updates the systemd_journal parser to support systemd versions up to 254.

## Description:
- fix header sizes
- add compact mode
- add support for zstd compression

~~TODO: add zstd to requirements.txt~~

**Related issue (if applicable):** fixes #4728

## Notes:
All contributions to Plaso undergo [code review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc).
This makes sure that the code has appropriate test coverage and conforms to the
[Plaso style guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
* [x] Automated checks (GitHub Actions, AppVeyor) pass
* [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
* [x] Reviewer assigned

### CODE REVIEW

**Comment by michael-ashburn** [comment]
[user-1000.journal.zip](https://github.com/log2timeline/plaso/files/12727510/user-1000.journal.zip)

**Comment by joachimmetz** [comment]
@michael-ashburn thanks for the changes

@sydp can you take a first pass over this PR

**Comment by joachimmetz** [comment]
@michael-ashburn which zstd dependency are you using?

**Comment by michael-ashburn** [comment]
https://pypi.org/project/zstd/. Checked with 1.5.5.1 and 1.5.2.4.

File: plaso/parsers/systemd_journal.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 from dfdatetime import posix_time as dfdatetime_posix_time
 
+import zstd
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:9]: move this import one line down to maintain import ordering

File: plaso/parsers/systemd_journal.py [hunk 2]
```diff
@@ -96,10 +101,14 @@ def _ParseDataObject(self, file_object, file_offset):
     Raises:
       ParseError: if the data object cannot be parsed.
     """
-    data_object_map = self._GetDataTypeMap('systemd_journal_data_object')
+    if not self._compact:
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:104]: I would avoid negative logic in this case.

File: plaso/parsers/systemd_journal.py [hunk 3]
```diff
@@ -127,14 +137,23 @@ def _ParseDataObject(self, file_object, file_offset):
 
       try:
         uncompressed_size = self._ReadStructureFromByteStream(
-            data, file_offset + 64, uncompressed_size_map)
+            data, file_offset + data_object_header_size, uncompressed_size_map)
       except (ValueError, errors.ParseError) as exception:
         raise errors.ParseError((
             'Unable to parse LZ4 uncompressed size at offset: 0x{0:08x} with '
-            'error: {1!s}').format(file_offset + 64, exception))
+            'error: {1!s}').format(file_offset + data_object_header_size,
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:144]: style nit: move `file_offset...` to the a new line with hanging indent

File: plaso/parsers/systemd_journal.py [hunk 4]
```diff
@@ -151,8 +170,12 @@ def _ParseEntryArrayObject(self, file_object, file_offset):
     Raises:
       ParseError: if the entry array object cannot be parsed.
     """
-    entry_array_object_map = self._GetDataTypeMap(
+    if not self._compact:
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:173]: as above, I would invert the logic here.

File: plaso/parsers/systemd_journal.py [hunk 5]
```diff
@@ -151,8 +170,12 @@ def _ParseEntryArrayObject(self, file_object, file_offset):
...
+      entry_array_object_map = self._GetDataTypeMap(
         'systemd_journal_entry_array_object')
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:175]: style nit: add 2 spaces

File: plaso/parsers/systemd_journal.py [hunk 6]
```diff
@@ -151,8 +170,12 @@ def _ParseEntryArrayObject(self, file_object, file_offset):
...
+    else:
+      entry_array_object_map = self._GetDataTypeMap(
+        'systemd_journal_entry_array_object_compact')
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:178]: style nit: add 2 spaces

File: plaso/parsers/systemd_journal.py [hunk 7]
```diff
@@ -246,7 +269,11 @@ def _ParseJournalEntry(self, file_object, file_offset):
     entry_object = self._ParseEntryObject(file_object, file_offset)
 
     # The data is read separately for performance reasons.
-    entry_item_map = self._GetDataTypeMap('systemd_journal_entry_item')
+    if not self._compact:
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:272]: as above, invert the logic here

File: plaso/parsers/systemd_journal.py [hunk 8]
```diff
@@ -246,7 +269,11 @@ def _ParseJournalEntry(self, file_object, file_offset):
...
+      entry_item_map = self._GetDataTypeMap('systemd_journal_entry_item')
+    else:
+      entry_item_map = self._GetDataTypeMap(
+        'systemd_journal_entry_item_compact')
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:276]: style nit: add 2 spaces

**Review by sydp — CHANGES_REQUESTED** [review]
Thanks for the PR.  

Could you also add the test data file and update the unit tests for this parser?

**Comment by sydp** [comment]
> https://pypi.org/project/zstd/. Checked with 1.5.5.1 and 1.5.2.4.

To add a new dependency I believe this will need to be added to l2tdevtools.

File: plaso/parsers/systemd_journal.py [hunk 9]
```diff
@@ -66,8 +68,8 @@ class SystemdJournalParser(
   _HEADER_SIZE_VERSION_187 = 224
   _HEADER_SIZE_VERSION_189 = 240
   _HEADER_SIZE_VERSION_246 = 256
-  _HEADER_SIZE_VERSION_252 = 268
-  _HEADER_SIZE_VERSION_254 = 276
+  _HEADER_SIZE_VERSION_252 = 264
```
Comment (sydp) [inline plaso/parsers/systemd_journal.py:71]: fwiw I also have a fix for this here - https://github.com/log2timeline/plaso/pull/4738

**Comment by joachimmetz** [comment]
> To add a new dependency I believe this will need to be added to l2tdevtools.

Will take a look when time permits

**Comment by joachimmetz** [comment]
@michael-ashburn can you allow maintainer pushes on your fork so I can push the necessary changes to add zstd

File: plaso/parsers/systemd_journal.py (hunk 9 above)
Comment (joachimmetz) [reply plaso/parsers/systemd_journal.py:71]: @sydp given #4738 is a very small PR I would recommend closing it in favor of this one

File: dependencies.ini [hunk 10]
```diff
@@ -478,3 +478,12 @@ minimum_version: 2.1.11
 pypi_name: pyzmq
 rpm_name: python3-zmq
 version_property: __version__
+
+[zstd]
+dpkg_name: python3-zstd
+l2tbinaries_macos_name: python3-zstd
```
Comment (joachimmetz) [inline dependencies.ini:484]: l2tbinaries_macos_name is not longer used and can be removed

**Review by joachimmetz — APPROVED** [review]
LGTM

**Pull request merged**